pulsar-broker/src/test/java/org/apache/pulsar/client/impl/KeySharedSubscriptionTest.java

```diff
@@ -0,0 +1,162 @@
+                    nameToId.computeIfAbsent(cons1,(k) -> new ArrayList<>())
+                            .add(msg.getMessageId());
+                    msgCount.incrementAndGet();
+                    if (canAcknowledgement.get()) {
+                        try {
+                            cons1.acknowledge(msg);
+                        } catch (PulsarClientException e) {
+                            throw new RuntimeException(e);
+                        }
+                    }
+                })
+                .subscribe();
+        @Cleanup
+        Consumer<byte[]> consumer2 = pulsarClient.newConsumer()
+                .topic(topic)
+                .subscriptionName("sub-1")
+                .subscriptionType(SubscriptionType.Key_Shared)
+                .messageListener((cons2, msg) -> {
+                    lastActiveTime.set(System.currentTimeMillis());
+                    nameToId.computeIfAbsent(cons2,(k) -> new ArrayList<>())
+                            .add(msg.getMessageId());
+                    msgCount.incrementAndGet();
+                    if (canAcknowledgement.get()) {
+                        try {
+                            cons2.acknowledge(msg);
+                        } catch (PulsarClientException e) {
+                            throw new RuntimeException(e);
+                        }
+                    }
+                })
+                .consumerName("con-2")
+                .subscribe();
+        @Cleanup
+        Consumer<byte[]> consumer3 = pulsarClient.newConsumer()
+                .topic(topic)
+                .subscriptionName("sub-1")
+                .subscriptionType(SubscriptionType.Key_Shared)
+                .messageListener((cons3, msg) -> {
+                    lastActiveTime.set(System.currentTimeMillis());
+                    nameToId.computeIfAbsent(cons3,(k) -> new ArrayList<>())
+                            .add(msg.getMessageId());
+                    msgCount.incrementAndGet();
+                    if (canAcknowledgement.get()) {
+                        try {
+                            cons3.acknowledge(msg);
+                        } catch (PulsarClientException e) {
+                            throw new RuntimeException(e);
+                        }
+                    }
+                })
+                .consumerName("con-3")
+                .subscribe();
+
+        @Cleanup
+        Producer<byte[]> producer = pulsarClient.newProducer()
+                .topic(topic)
+                .enableBatching(true)
+                .create();
+        for (int i = 0; i < 50000; i++) {
+            producer.sendAsync(UUID.randomUUID().toString().getBytes(StandardCharsets.UTF_8));
+        }
+
+        while (true) {
+            long differentTime = System.currentTimeMillis() - lastActiveTime.get();
+            if (differentTime > TimeUnit.SECONDS.toMillis(20)) {
+                break;
+            }
+        }
```

Comment:
Could you please help add some description? I'm sure how the code will affect the test.

Reply:
fixed.

Reply:
I see. Can we use Awaitiblity here? 
We don't have any poll interval for the while loop, it looks like busy loop which will waste CPU.

Reply:
good idea.